strconv.ParseInt panic when reading a float value into an `int`

### Terraform Version
```
2019/05/10 10:53:41 [INFO] Terraform version: 0.12.0 rc1
2019/05/10 10:53:41 [INFO] Go runtime version: go1.12.1
+ provider.google 2.7.0 (unreleased)
```

### Terraform Configuration Files
```hcl
resource "google_compute_backend_service" "go_clicks" {
  name          = "go-clicks"
  description   = "backend service for go-clicks from different clusters"
  health_checks = ["fake_self_link_for_github_issue"]

  load_balancing_scheme = "EXTERNAL"
  protocol = "HTTP"

  timeout_sec = 30.02

  backend {
    max_rate = 99999
    balancing_mode = "RATE"
  }
}
```

### Debug Output
https://gist.github.com/chrisst/1a500d9a40dbbaab2d1b540ea4a69a51

### Crash Output
https://gist.github.com/chrisst/a7100eb4dbe8ecf581c4157f03ebf6b1

### Expected Behavior
Terraform should throw a type validation error rather than panicking 

### Actual Behavior
panic: Error reading level config: strconv.ParseInt: parsing "30.02": invalid syntax

### Steps to Reproduce
1. Set a float value for a field defined as an integer
2. Terraform plan

### Additional Context
I confirmed this bug exists with Terraform 0.11 and a provider version running a pre 0.12 sdk as well (google-provider 2.4.1).

### References
May be related to https://github.com/hashicorp/terraform/issues/18121 however this doesn't require any migrations in order to repro. 
Could be the root cause of https://github.com/terraform-providers/terraform-provider-google/issues/3590

--- a/builtin/providers/test/resource.go
+++ b/builtin/providers/test/resource.go
@@ -153,6 +153,10 @@ func testResource() *schema.Resource {
 				Optional:    true,
 				Description: "do not set in config",
 			},
+			"int": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
 		},
 	}
 }

--- a/builtin/providers/test/resource_test.go
+++ b/builtin/providers/test/resource_test.go
@@ -1029,3 +1029,24 @@ resource "test_resource" "foo" {
 		},
 	})
 }
+
+func TestResource_floatInIntAttr(t *testing.T) {
+	resource.UnitTest(t, resource.TestCase{
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckResourceDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: strings.TrimSpace(`
+resource "test_resource" "foo" {
+	required = "yep"
+	required_map = {
+	    key = "value"
+	}
+	int = 40.2
+}
+				`),
+				ExpectError: regexp.MustCompile(`must be a whole number, got 40.2`),
+			},
+		},
+	})
+}

--- a/helper/schema/schema.go
+++ b/helper/schema/schema.go
@@ -1731,12 +1731,25 @@ func (m schemaMap) validatePrimitive(
 		}
 		decoded = n
 	case TypeInt:
-		// Verify that we can parse this as an int
-		var n int
-		if err := mapstructure.WeakDecode(raw, &n); err != nil {
-			return nil, []error{fmt.Errorf("%s: %s", k, err)}
+		switch {
+		case isProto5():
+			// We need to verify the type precisely, because WeakDecode will
+			// decode a float as an integer.
+
+			// the config shims only use int for integral number values
+			if v, ok := raw.(int); ok {
+				decoded = v
+			} else {
+				return nil, []error{fmt.Errorf("%s: must be a whole number, got %v", k, raw)}
+			}
+		default:
+			// Verify that we can parse this as an int
+			var n int
+			if err := mapstructure.WeakDecode(raw, &n); err != nil {
+				return nil, []error{fmt.Errorf("%s: %s", k, err)}
+			}
+			decoded = n
 		}
-		decoded = n
 	case TypeFloat:
 		// Verify that we can parse this as an int
 		var n float64

